File: PandorasBox/Ahorn/entities/propellerbox.jl
module PandorasBoxPropellerBox

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/propellerBox" PropellerBox(x::Integer, y::Integer, texture::String="default", flashUseColor::String="3F437C", flashChargeColor::String="5A1C1C", maxCharges::Integer=3)

const placements = Ahorn.PlacementDict(
    "Propeller Box (Pandora's Box)" => Ahorn.EntityPlacement(
        PropellerBox
    )
)

function getSprite(entity::PropellerBox)
    texture = get(entity, "texture", "default")

    return "objects/pandorasBox/propellerBox/$texture/default_charges00"
end


function Ahorn.selection(entity::PropellerBox)
    x, y = Ahorn.position(entity)
    sprite = getSprite(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y, jx=0.5, jy=1.0)
end

Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::PropellerBox, room::Maple.Room) = Ahorn.drawSprite(ctx, getSprite(entity), 0, 0, jx=0.5, jy=1.0)

end

File: PandorasBox/Loenn/entities/propeller_box.lua
local propellerBox = {}

local glidingModes = {
    {"Always", "Always"},
    {"After Using Charge", "AfterUse"},
    {"Never", "Never"}
}

propellerBox.name = "pandorasBox/propellerBox"
propellerBox.depth = 50
propellerBox.fieldInformation = {
    maxCharges = {
        fieldType = "integer",
    },
    flashUseColor = {
        fieldType = "color",
        allowXNAColors = true,
    },
    flashChargeColor = {
        fieldType = "color",
        allowXNAColors = true,
    },
    glideMode = {
        options = glidingModes,
        editable = false
    }
}
propellerBox.placements = {
    name = "propeller_box",
    data = {
        texture = "default",
        flashUseColor = "3F437C",
        flashChargeColor = "5A1C1C",
        maxCharges = 3,
        rechargeOnGround = true,
        glideMode = "AfterUse",
    }
}

propellerBox.justification = {0.5, 1.0}

function propellerBox.texture(room, entity)
    local texture = entity.texture or "default"

    return string.format("objects/pandorasBox/propellerBox/%s/default_charges00", texture)
end

return propellerBox